bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/profiles/ModbusProfileFactory.java

```diff
@@ -0,0 +1,114 @@
+ */
+package org.openhab.binding.modbus.internal.profiles;
+
+import static org.openhab.binding.modbus.internal.profiles.ModbusProfiles.*;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.i18n.LocalizedKey;
+import org.openhab.core.thing.UID;
+import org.openhab.core.thing.profiles.Profile;
+import org.openhab.core.thing.profiles.ProfileCallback;
+import org.openhab.core.thing.profiles.ProfileContext;
+import org.openhab.core.thing.profiles.ProfileFactory;
+import org.openhab.core.thing.profiles.ProfileType;
+import org.openhab.core.thing.profiles.ProfileTypeProvider;
+import org.openhab.core.thing.profiles.ProfileTypeUID;
+import org.openhab.core.thing.profiles.i18n.ProfileTypeI18nLocalizationService;
+import org.openhab.core.util.BundleResolver;
+import org.osgi.framework.Bundle;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * A factory and advisor for modbus profiles.
+ *
+ *
+ * @author Sami Salonen - Initial contribution
+ */
+@Component(service = { ProfileFactory.class, ProfileTypeProvider.class })
+@NonNullByDefault
+public class ModbusProfileFactory implements ProfileFactory, ProfileTypeProvider {
+
+    private static final Set<ProfileType> SUPPORTED_PROFILE_TYPES = Set.of(GAIN_OFFSET_TYPE);
+
+    private static final Set<ProfileTypeUID> SUPPORTED_PROFILE_TYPE_UIDS = Set.of(GAIN_OFFSET);
+
+    private final Map<LocalizedKey, ProfileType> localizedProfileTypeCache = new ConcurrentHashMap<>();
+
+    private final ProfileTypeI18nLocalizationService profileTypeI18nLocalizationService;
+    private final Bundle bundle;
+
+    @Activate
+    public ModbusProfileFactory(final @Reference ProfileTypeI18nLocalizationService profileTypeI18nLocalizationService,
+            final @Reference BundleResolver bundleResolver) {
+        this.profileTypeI18nLocalizationService = profileTypeI18nLocalizationService;
+        this.bundle = bundleResolver.resolveBundle(ModbusProfileFactory.class);
+    }
+
+    @Override
+    public @Nullable Profile createProfile(ProfileTypeUID profileTypeUID, ProfileCallback callback,
+            ProfileContext context) {
+        if (GAIN_OFFSET.equals(profileTypeUID)) {
+            return new ModbusGainOffsetProfile(callback, context);
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public Collection<ProfileType> getProfileTypes(@Nullable Locale locale) {
+        return Collections.unmodifiableList(SUPPORTED_PROFILE_TYPES.stream()
+                .map(p -> createLocalizedProfileType(p, locale)).collect(Collectors.toList()));
```

Comment:
What's the reason for creating a i18n Profile? If I see correctly you don't use any i18n in the Profile.

Reply:
I have followed 'best practice' established by system profiles: `SystemProfileFactory` in https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/profiles/SystemProfileFactory.java

`ProfileTypeRegistryImpl` in https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/profiles/ProfileTypeRegistryImpl.java will be calling this method.

But your are right, I am not providing any localization at this stage.

Reply:
Are you planning to add i18n? I personally don't like unnecessary code...

Reply:
Makes sense, agreed.

Do I need to implement ProfileTypeProvider or can I remove it alltogether?

Reply:
Frankly, I don't know. I implemented it, but I simply copied code at that time I guess: https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.lcn/src/main/java/org/openhab/binding/lcn/internal/LcnProfileFactory.java

Reply:
This resulted in a nice cleanup: https://github.com/openhab/openhab-addons/pull/9980/commits/3845e87ca613acb301442f720252d3542565a3d9

Reply:
Just for curiosity, did you try to remove the `ProfileTypeProvider`, too?

Reply:
Did not try it 🙈